### PR TITLE
Add StyleCop.Analyzers and enable some basic suggestions

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Adding `StyleCop.Analyzers` to help us enforce some good practices. We should be able to apply the `SA1122` rule globally using the `dotnet format` to replace all `""` with `string.Empty`.

Also, at some point I think almost all the rules that are defined as `Warning` in the link below should be defined at least as `suggestion` in OC.

https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/rulesets/StyleCopAnalyzersDefault.ruleset

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -80,3 +80,207 @@ dotnet_naming_rule.private_const_fields_ignore_camel_case_leading_underscore.sev
...
+dotnet_diagnostic.SA1108.severity = none			# Block statements should not contain embedded comments.
+dotnet_diagnostic.SA1109.severity = none			# Block statements should not contain embedded regions.
+dotnet_diagnostic.SA1110.severity = none			# Opening parenthesis or bracket should be on declaration line.
+dotnet_diagnostic.SA1111.severity = none			# Closing parenthesis should be on line of last parameter.
+dotnet_diagnostic.SA1112.severity = none			# Closing parenthesis should be on line of opening parenthesis.
+dotnet_diagnostic.SA1113.severity = none			# Comma should be on the same line as previous parameter.
+dotnet_diagnostic.SA1114.severity = none			# Parameter list should follow declaration.
+dotnet_diagnostic.SA1115.severity = none			# Parameter should follow comma .
+dotnet_diagnostic.SA1116.severity = none			# Split parameters should start on line after declaration.
+dotnet_diagnostic.SA1117.severity = suggestion		# Parameters should be on same line or separate lines.
```
Comment (jtkech) [inline .editorconfig:136]: Maybe not use tabs for the indentation.

Do we need rules for the .editorconfig itself ;)

File: .editorconfig [hunk 2]
```diff
@@ -80,3 +80,207 @@ dotnet_naming_rule.private_const_fields_ignore_camel_case_leading_underscore.sev
...
+dotnet_diagnostic.SA1504.severity = suggestion		# All accessors should be single-line or multi-line.
+dotnet_diagnostic.SA1505.severity = suggestion		# A closing brace should not be preceded by a blank line.
+dotnet_diagnostic.SA1506.severity = suggestion		# Element documentation headers should not be followed by blank line.
+dotnet_diagnostic.SA1507.severity = none			# Code should not contain multiple blank lines in a row.
+dotnet_diagnostic.SA1508.severity = none			# Closing braces should not be preceded by blank line.
+dotnet_diagnostic.SA1509.severity = none			# Opening braces should not be preceded by blank line.
+dotnet_diagnostic.SA1510.severity = none			# Chained statement blocks should not be preceded by blank line.
+dotnet_diagnostic.SA1511.severity = none			# While-do footer should not be preceded by blank line.
+dotnet_diagnostic.SA1512.severity = none			# Single-line comments should not be followed by blank line.
+dotnet_diagnostic.SA1513.severity = none			# Closing brace should be followed by blank line.
```
Comment (jtkech) [inline .editorconfig:225]: Didn't find a rule to add a new line after a closing brace ending a code block(s), e.g. after a if/while/for blocks.

**Comment by jtkech** [comment]
Related to #12929 and #7950

Yes, I thought about this but the first goal was just to follow our current editor rules and the default analyzers, because it was already implying many updates.

Updated Analyzers and new ones will come with net8.0, trying dotnet 8.0 I could saw new code analysis suggestions.

Otherwise looks good, we will need more time to check the rules we want, at least the ones that we already follow implicitly, didn't try to define them without having to use `StyleCop`.

File: .editorconfig [hunk 3]
```diff
@@ -80,3 +80,207 @@ dotnet_naming_rule.private_const_fields_ignore_camel_case_leading_underscore.sev
...
+# Rules are listed at https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/rulesets/StyleCopAnalyzersDefault.ruleset
+
+# StyleCop: Spacing Rules
+[*.cs]
+dotnet_diagnostic.SA1000.severity = none 			# Keywords should be spaced correctly.
+dotnet_diagnostic.SA1001.severity = none			# Commas should be spaced correctly.
+dotnet_diagnostic.SA1002.severity = none			# Semicolons should be spaced correctly.
+dotnet_diagnostic.SA1003.severity = none			# Symbols should be spaced correctly.
+dotnet_diagnostic.SA1004.severity = none			# Documentation lines should begin with single space.
+dotnet_diagnostic.SA1005.severity = none			# Single line comment should begin with a space.
```
Comment (jtkech) [inline .editorconfig:93]: This one is good.

Didn't find one about starting with an uppercase and ending with a dot.

File: .editorconfig [hunk 4]
```diff
@@ -80,3 +80,207 @@ dotnet_naming_rule.private_const_fields_ignore_camel_case_leading_underscore.sev
...
+dotnet_diagnostic.SA1506.severity = suggestion		# Element documentation headers should not be followed by blank line.
+dotnet_diagnostic.SA1507.severity = none			# Code should not contain multiple blank lines in a row.
+dotnet_diagnostic.SA1508.severity = none			# Closing braces should not be preceded by blank line.
+dotnet_diagnostic.SA1509.severity = none			# Opening braces should not be preceded by blank line.
+dotnet_diagnostic.SA1510.severity = none			# Chained statement blocks should not be preceded by blank line.
+dotnet_diagnostic.SA1511.severity = none			# While-do footer should not be preceded by blank line.
+dotnet_diagnostic.SA1512.severity = none			# Single-line comments should not be followed by blank line.
+dotnet_diagnostic.SA1513.severity = none			# Closing brace should be followed by blank line.
+dotnet_diagnostic.SA1514.severity = none			# Element documentation header should be preceded by blank line.
+dotnet_diagnostic.SA1515.severity = none			# Single-line comment should be preceded by blank line.
```
Comment (jtkech) [inline .editorconfig:227]: This one is good

**Comment by MikeAlhayek** [comment]
@jtkech I cleaned up the `.editorconfig` file. I also enabled few rules that we currently follow in OC. Once we merge this PR, we can then enable and apply globally other rules.

The idea with this PR here is to include the StyleCop analyzer and then use it in other PR along with applying rules.

**Comment by MikeAlhayek** [comment]
@jtkech you recently did lots of clean up. Should we merge this one?

**Comment by jtkech** [comment]
Refer to https://github.com/OrchardCMS/OrchardCore/pull/14074#issuecomment-1664810433

> @jtkech you recently did lots of clean up.

Yes, to follow our current rules and default analyzers, already implying many updates.

> Should we merge this one?

I would like too to have at least the rules that we already follow implicitly, so I'm not against.

Just need to ask @sebastienros if he agrees to use StyleCop, knowing that:

- Updated Analyzers and new ones will come with net8.0.

- Didn't try to define these rules without using StyleCop, maybe not possible.

- We may need a little more time to check the rules we want.

Yes, if we use StyleCop I will take the time to try some rules, not sure about `String.Empty` (but not against), "" does the same and there are many, we could add rules around the usage of braces ...

For info about our current rules, we have remaining violations, it looks like that when we run the code analysis it doesn't check the whole files, but I think we fixed most of the violations.

File: .editorconfig [hunk 5]
```diff
@@ -80,3 +80,207 @@ dotnet_naming_rule.private_const_fields_ignore_camel_case_leading_underscore.sev
...
+dotnet_diagnostic.SA1107.severity = none        	# Code should not contain multiple statements on one line.
+dotnet_diagnostic.SA1108.severity = none        	# Block statements should not contain embedded comments.
+dotnet_diagnostic.SA1109.severity = none        	# Block statements should not contain embedded regions.
+dotnet_diagnostic.SA1110.severity = none        	# Opening parenthesis or bracket should be on declaration line.
+dotnet_diagnostic.SA1111.severity = none        	# Closing parenthesis should be on line of last parameter.
+dotnet_diagnostic.SA1112.severity = none        	# Closing parenthesis should be on line of opening parenthesis.
+dotnet_diagnostic.SA1113.severity = none        	# Comma should be on the same line as previous parameter.
+dotnet_diagnostic.SA1114.severity = none        	# Parameter list should follow declaration.
+dotnet_diagnostic.SA1115.severity = none        	# Parameter should follow comma .
+dotnet_diagnostic.SA1116.severity = none        	# Split parameters should start on line after declaration.
```
Comment (jtkech) [inline .editorconfig:135]: Looks like there are some remaining tab indentations

File: .editorconfig [hunk 6]
```diff
@@ -80,3 +80,207 @@ dotnet_naming_rule.private_const_fields_ignore_camel_case_leading_underscore.sev
...
+# StyleCop: Layout Rules
+dotnet_diagnostic.SA1500.severity = none            # Braces for multi-line statements should not share line.
+dotnet_diagnostic.SA1501.severity = none            # Statement should not be on a single line.
+dotnet_diagnostic.SA1502.severity = none            # Element should not be on a single line.
+dotnet_diagnostic.SA1503.severity = none            # Braces should not be omitted.
+dotnet_diagnostic.SA1504.severity = none            # All accessors should be single-line or multi-line.
+dotnet_diagnostic.SA1505.severity = none            # A closing brace should not be preceded by a blank line.
+dotnet_diagnostic.SA1506.severity = none            # Element documentation headers should not be followed by blank line.
+dotnet_diagnostic.SA1507.severity = none            # Code should not contain multiple blank lines in a row.
+dotnet_diagnostic.SA1508.severity = none            # Closing braces should not be preceded by blank line.
```
Comment (jtkech) [inline .editorconfig:220]: Yes, we could suggest this one

File: .editorconfig [hunk 7]
```diff
@@ -80,3 +80,207 @@ dotnet_naming_rule.private_const_fields_ignore_camel_case_leading_underscore.sev
...
+dotnet_diagnostic.SA1509.severity = none            # Opening braces should not be preceded by blank line.
+dotnet_diagnostic.SA1510.severity = none            # Chained statement blocks should not be preceded by blank line.
+dotnet_diagnostic.SA1511.severity = none            # While-do footer should not be preceded by blank line.
+dotnet_diagnostic.SA1512.severity = none            # Single-line comments should not be followed by blank line.
+dotnet_diagnostic.SA1513.severity = none            # Closing brace should be followed by blank line.
```
Comment (jtkech) [inline .editorconfig:225]: Idem, would need a little more time to check all of them ;)

File: src/OrchardCore.Build/Dependencies.props [hunk 8]
```diff
@@ -64,5 +64,6 @@
     <PackageManagement Include="YesSql.Filters.Abstractions" Version="3.3.0" />
     <PackageManagement Include="YesSql.Filters.Query" Version="3.3.0" />
     <PackageManagement Include="ZString" Version="2.5.0" />
+    <PackageManagement Include="StyleCop.Analyzers" Version="1.1.118" />
```
Comment (jtkech) [inline src/OrchardCore.Build/Dependencies.props:67]: Depedencies are defined in alphabetic order

Comment (MikeAlhayek) [reply src/OrchardCore.Build/Dependencies.props]: Done.

File: .editorconfig (hunk 6 above)
Comment (MikeAlhayek) [reply .editorconfig:220]: @jtkech 
yes we could suggest more than just this one. But, I think in this PR, we should merge it as is if agreed. Once we merge it, we can use follow up PR to enable and apply which every rules we want. this way, it's easy to clean up the project using smaller PRs with one/two rules at a time.

**Comment by jtkech** [comment]
> yes we could suggest more than just this one. But, I think in this PR, we should merge it as is if agreed. Once we merge it, we can use follow up PR to enable and apply which every rules we want. this way, it's easy to clean up the project using smaller PRs with one/two rules at a time.

Okay then, so I think it is too soon to suggest `String.Empty`, I suggest to remove this rule.

Then I will approve and merge, unless you prefer to wait for @sebastienros, if he prefers to not use StyleCop, hmm but anyway easy to revert.

**Comment by MikeAlhayek** [comment]
@jtkech okay. I switched SA1122 to none for now. We always prefer `String.Empty` over `""` which is why it was suggested. But yes, that can also be part of a separate PR

**Review by jtkech — APPROVED** [review]

**Comment by jtkech** [comment]
Yes I agree but for now to not add 500 sugestions in the current code.

**Pull request merged**